File: backend/app/schemas/place.py
from datetime import time, datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.enums import PlaceStatus, AgeGroup, PlaceSortBy
from app.schemas.category import CategoryResponse
from app.schemas.interest_tag import InterestTagResponse


class PlaceImageBase(BaseModel):
    img_url: str
    caption: str | None = None
    is_primary: bool = False


class PlaceImageCreate(PlaceImageBase):
    pass


class PlaceImageResponse(PlaceImageBase):
    id: int

    class Config:
        from_attributes = True


class PlaceTagCreate(BaseModel):
    tag_id: int
    relevance: Decimal = Field(default=Decimal('1.0'), ge=0, le=1)


class PlaceTagResponse(BaseModel):
    relevance: Decimal
    tag: InterestTagResponse

    class Config:
        from_attributes = True


class PlaceAgeGroupBase(BaseModel):
    age_group: AgeGroup
    suitability: int = Field(default=3, ge=1, le=5)


class PlaceAgeGroupCreate(PlaceAgeGroupBase):
    pass


class PlaceAgeGroupResponse(PlaceAgeGroupBase):
    class Config:
        from_attributes = True


class PlaceBase(BaseModel):
    name: str
    description: str
    address: str
    ward: str
    link_google_map: str
    phone: str | None = None
    website: str | None = None
    price_min: Decimal = Decimal("0")
    price_max: Decimal = Decimal("0")
    opening_time: time
    closing_time: time
    open_days: str


class PlaceCreate(PlaceBase):
    is_featured: bool = False
    status: PlaceStatus = PlaceStatus.ACTIVE
    category_ids: list[int] = []
    tags: list[PlaceTagCreate] = []
    age_groups: list[PlaceAgeGroupCreate] = []
    images: list[PlaceImageCreate] = []


class PlaceUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    address: str | None = None
    ward: str | None = None
    link_google_map: str | None = None
    phone: str | None = Field(default=None, min_length=10)
    website: str | None = None
    price_min: Decimal | None = None
    price_max: Decimal | None = None
    opening_time: time | None = None
    closing_time: time | None = None
    open_days: str | None = None
    is_featured: bool | None = None
    status: PlaceStatus | None = None
    category_ids: list[int] | None = None
    tags: list[PlaceTagCreate] | None = None
    age_groups: list[PlaceAgeGroupCreate] | None = None


class PlaceResponse(PlaceBase):
    id: int
    average_rating: Decimal
    total_reviews: int
    total_views: int
    is_featured: bool
    status: PlaceStatus
    created_by: int | None = None
    created_at: datetime
    updated_at: datetime

    categories: list[CategoryResponse] = []
    tags: list[PlaceTagResponse] = []
    age_groups: list[PlaceAgeGroupResponse] = []
    images: list[PlaceImageResponse] = []

    class Config:
        from_attributes = True


class PlaceStatusUpdate(BaseModel):
    status: PlaceStatus


class PlaceFeaturedUpdate(BaseModel):
    is_featured: bool


class PlaceSummaryResponse(BaseModel):
    id: int
    name: str
    address: str
    ward: str
    average_rating: Decimal
    total_reviews: int
    total_views: int
    is_featured: bool
    status: PlaceStatus
    primary_image: str | None = None
    opening_time: time
    closing_time: time

    class Config:
        from_attributes = True


class PlaceSearchResponse(BaseModel):
    total: int
    skip: int
    limit: int
    sort_by: PlaceSortBy
    items: list[PlaceSummaryResponse] = []


class WardCountResponse(BaseModel):
    ward: str
    total: int

File: admin-web/src/types/place.ts
import { CategoryResponse } from "./category";
import { InterestTagResponse } from "./tag";

export type PlaceStatus = 'ACTIVE' | 'CLOSED' | 'PENDING';
export type AgeGroup = 'CHILDREN' | 'TEENAGER' | 'YOUNG_ADULT' | 'ADULT' | 'MIDDLE_AGE' | 'SENIOR';
export interface PlaceImageCreate {
    img_url: string;
    caption?: string | null;
    is_primary?: boolean;
}
export interface PlaceImageResponse extends PlaceImageCreate {
    id: number;
}
export interface PlaceTagCreate {
    tag_id: number;
    relevance?: string;
}
export interface PlaceTagResponse {
    relevance: string;
    tag: InterestTagResponse;
}
export interface PlaceAgeGroupCreate {
    age_group: AgeGroup;
    suitability?: number;
}
export interface PlaceAgeGroupResponse {
    age_group: AgeGroup;
    suitability: number;
}
export interface PlaceBase {
    name: string; description: string; address: string; ward: string;
    link_google_map: string; phone: string | null; website?: string | null;
    price_min: string; price_max: string;
    opening_time: string; closing_time: string; open_days: string;
}
export interface PlaceCreate extends PlaceBase {
    is_featured?: boolean; status?: PlaceStatus;
    category_ids: number[]; tags?: PlaceTagCreate[]; age_groups?: PlaceAgeGroupCreate[];
    images?: PlaceImageCreate[];
}

export interface PlaceUpdate {
    name?: string; description?: string; address?: string; ward?: string;
    link_google_map?: string; phone?: string | null; website?: string | null;
    price_min?: string; price_max?: string;
    opening_time?: string; closing_time?: string; open_days?: string;
    category_ids?: number[]; tags?: PlaceTagCreate[]; age_groups?: PlaceAgeGroupCreate[];
}

export interface PlaceResponse extends PlaceBase {
    id: number; average_rating: string; total_reviews: number; total_views: number;
    is_featured: boolean; status: PlaceStatus; created_by: number | null;
    created_at: string; updated_at: string;
    categories: CategoryResponse[]; tags: PlaceTagResponse[];
    age_groups: PlaceAgeGroupResponse[]; images: PlaceImageResponse[];
}

export interface PlaceSummaryResponse {
    id: number; name: string; ward: string; average_rating: string;
    total_reviews: number; total_views: number; is_featured: boolean;
    status: PlaceStatus; primary_image: string | null;
}

export interface PlaceStatusUpdate { status: PlaceStatus }
export interface PlaceFeaturedUpdate { is_featured: boolean; }
export interface WardCountResponse { ward: string; total: number; }
export const AGE_GROUP_LABELS: Record<AgeGroup, string> = {
    CHILDREN:'Trẻ em', TEENAGER:'Thiếu niên', YOUNG_ADULT:'Thanh niên',
    ADULT:'Người lớn', MIDDLE_AGE:'Trung niên', SENIOR:'Cao tuổi'
}
export const PLACE_STATUS_LABELS: Record<PlaceStatus, string> ={
    ACTIVE:'Hoạt động', PENDING:'Chờ duyệt', CLOSED:'Đã đóng'
};

